crates/bevy_ecs/src/world/entity_borrow.rs

```diff
@@ -0,0 +1,447 @@
+    /// detection in custom runtimes.
+    ///
+    /// **You should prefer to use the typed API [`EntityMut::get_change_ticks`] where possible and only
+    /// use this in cases where the actual component types are not known at
+    /// compile time.**
+    #[inline]
+    pub fn get_change_ticks_by_id(&self, component_id: ComponentId) -> Option<ComponentTicks> {
+        self.as_readonly().get_change_ticks_by_id(component_id)
+    }
+
+    /// Gets the component of the given [`ComponentId`] from the entity.
+    ///
+    /// **You should prefer to use the typed API [`EntityMut::get`] where possible and only
+    /// use this in cases where the actual component types are not known at
+    /// compile time.**
+    ///
+    /// Unlike [`EntityBorrowMut::get`], this returns a raw pointer to the component,
+    /// which is only valid while the [`EntityBorrowMut`] is alive.
+    #[inline]
+    pub fn get_by_id(&self, component_id: ComponentId) -> Option<Ptr<'_>> {
+        self.as_readonly().get_by_id(component_id)
+    }
+
+    /// Gets a [`MutUntyped`] of the component of the given [`ComponentId`] from the entity.
+    ///
+    /// **You should prefer to use the typed API [`EntityMut::get_mut`] where possible and only
+    /// use this in cases where the actual component types are not known at
+    /// compile time.**
+    ///
+    /// Unlike [`EntityMut::get_mut`], this returns a raw pointer to the component,
+    /// which is only valid while the [`EntityMut`] is alive.
+    #[inline]
+    pub fn get_mut_by_id(&mut self, component_id: ComponentId) -> Option<MutUntyped<'_>> {
+        // SAFETY:
+        // - `&mut self` ensures that no references exist to this entity's components.
+        // - `as_unsafe_world_cell` gives mutable permission for all components on this entity
+        unsafe { self.0.get_mut_by_id(component_id) }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{self as bevy_ecs, prelude::*, system::assert_is_system};
+
+    #[derive(Component)]
+    struct A;
+
+    #[test]
+    fn disjoint_access() {
+        fn disjoint_readonly(
+            _: Query<EntityBorrow, With<A>>,
+            _: Query<EntityBorrowMut, Without<A>>,
+        ) {
+        }
+
+        fn disjoint_mutable(
+            _: Query<EntityBorrowMut, With<A>>,
+            _: Query<EntityBorrowMut, Without<A>>,
+        ) {
+        }
+
+        assert_is_system(disjoint_readonly);
+        assert_is_system(disjoint_mutable);
+    }
+
+    #[test]
+    fn borrow_compatible() {
+        fn borrow_system(_: Query<(EntityBorrow, &A)>, _: Query<&A>) {}
+
+        assert_is_system(borrow_system);
+    }
+
+    #[test]
+    fn borrow_mut_compatible() {
+        fn borrow_mut_system(_: Query<(Entity, EntityBorrowMut)>) {}
+
+        assert_is_system(borrow_mut_system);
+    }
+
+    #[test]
+    #[should_panic]
+    fn borrow_mut_incompatible1() {
```

Comment:
These tests should have better names, even if they're more verbose. Maybe `borrow_mut_incompatible_with_read_only_component` and so on?